### DISCUSSION TITLE
Possible API improvement for `IdentityInput`

### DISCUSSION THREAD

**Discussion by Dhaiwat10** [FuelLabs/fuels-ts#1844] [Ideas]
Take this Sway function signature for example:

```sw
fn new_game(player: Identity);
```

If you try to call this function via the TS SDK, it would look something like this:

```ts
// for an `Address`
const wallet = Wallet.generate({});

await contract.functions.new_game({ Address: { value: wallet.address.toB256() });

// for a `Contract`
await contract.functions.new_game({ ContractId: { value: otherContract.id.toB256() });
```

It works, but this takes 3-4 attempts to get right and involves some mental gymnastics. I was eventually able to find out how to make this work because I have been working on the SDK for more than a year now. Most new users would find this daunting.

---

I feel like the API should be simplified to something like:

```ts
await contract.functions.new_game({ value: wallet.address }); // `Address`

await contract.functions.new_game({ value: contract.id, type: 'contractId' }); // `ContractId`
```

---

Any comments & ideas welcome!

**Comment by petertonysmith94** [comment]
💯 Love this!

---

Would suggest keeping the `ContractId` as the type value:
```ts
await contract.functions.new_game({ value: contract.id, type: 'ContractId' });
```

I assume we'll default to an `Address` type and the explicit form would be valid?
```ts
await contract.functions.new_game({ value: contract.address, type: 'Address' });
```

**Comment by arboleya** [comment]
I am lacking proper context here, but ideally, this would be:

```ts
await contract.functions.new_game(value);
```

This would feel more intuitive and standard (like other usages) since the Sway parameter is a single value:

```rust
fn new_game(player: Identity);
```

**Reply by Dhaiwat10** [reply]
@arboleya that wouldn't be possible since we have to accept both contract IDs and 'normal' addresses, and we need the user to notify the SDK about this in some way. This is why I think what @petertonysmith94 suggested might be the best option:

https://github.com/FuelLabs/fuels-ts/discussions/1844#discussioncomment-8743655

**Reply by arboleya** [reply]
Couldn't the SDK convert things automatically based on the ABI?

A simple string-based usage still sounds straightforward to me:
 - https://github.com/FuelLabs/fuels-ts/issues/1584

**Reply by arboleya** [reply]
FYI - new docs are incoming via https://github.com/FuelLabs/fuels-ts/pull/1870.

**Reply by petertonysmith94** [reply]
Does #1870 close this discussion then?

**Reply by arboleya** [reply]
Not necessarily.

IIUC, that one is _only_ adding docs while this one is considering a possible API change/simplification.

**Reply by Dhaiwat10** [reply]
@arboleya how would you do that via the ABI? 🤔 The ABI will only be able to tell us the fact that the given function accepts `Identity` - which can be an `Address` or `ContractId`. It is still upto the user to decide what they want to pass in as the input. I'm not following you here

**Reply by arboleya** [reply]
I see. Things are identical from the ABI's perspective.

I'm just trying to collaborate and cross the idea with previous conversations.

**Comment by nedsalk** [comment]
Great idea! A solution that'd make the `type` property obsolete could be the following:
```ts
type Identity = {address: string, contractId?: never} | {address?: never; contractId: string}

await contract.functions.new_game({ address: contract.address });
await contract.functions.new_game({ contractId: contract.id });
```
The `Identity` type allows only one of the properties to be defined; if somebody tries to assign values to both, it'll give them a compile-time error. I wrote a playground example [here](https://www.typescriptlang.org/play?#code/C4TwDgpgBAkgJhAdsAlqKBeKBvAhnOAJwgGcSAuKE4QlRAcwBooBjAe2UNxeHgH5KiCADcIhAL5QAPjnxFSJAVCGjCAblYca3XnErVaDcQChjCFgBtcxKADMAroh4oOyiAHcA+vVwBbCAAUYFYgYpTwSKigAJSUwmwocGqm7IjUUD7+AIyYbl6ZgXgExGSUAES+IAC0ciUkZeLRxqnpBQBMuUL5foWp2jzw5ZVVfVw8VYkNTc0c6YhsAOI9AGJshACabPadHt49AUXypVAV1bUKZcyjOoMn1+OTjcZAA).

**Reply by danielbate** [reply]
+1 for this implementation. We had this [forum post ](https://forum.fuel.network/t/fuelerror-only-one-field-must-be-provided/4550) where the user had a similar problem. IMO if you know the function takes an enum value it can be deduced from the types, but we can do more and I believe the above does that.

**Comment by arboleya** [comment]
Thanks, everyone.

I created an issue for it:
 - https://github.com/FuelLabs/fuels-ts/issues/1891
